MemoryMeter.c

```diff
@@ -36,7 +37,12 @@ static void MemoryMeter_updateValues(Meter* this) {
    /* Do not print available memory in bar mode */
    this->curItems = 4;
 
-   written = Meter_humanUnit(buffer, this->values[0], size);
+   /* Include shared in basic readout */
+   simple_used = this->values[0];
+   if (!isnan(this->values[1]))
```

Comment:
I would prefer this:
```suggestion
   if (this->values[1] > 0.0)
```
(With IEEE 754 semantics, NaN evaluate to `false` in all comparisons except `!=`.)

Reply:
True, but simple `>` on NaNs generates a floating point exception (`==`/`!=` doesn't).  At best that sets a sticky flag, at worst it actually traps and signals.

 `if (isgreater(this->values[1], 0.0))` would be appropriate. That acknowledges you're prepared for the "unordered" result and doesn't generate the exception.

@BenBE thoughts? Neat as that is, I kind of feel the explicit NaN test is more human-friendly.

Reply:
@kjbracey Good point. I never thought of floating point exceptions before (I thought the exceptions won't matter when I wrote the color graph code). But I feel awkward when I am forced to use C99-specific, unreadable floating point macros, just to avoid exceptions anyway.

I always believed `this->values[1] > 0.0` is more straightforward than `isgreater(this->values[1], 0.0)`, and the configure script should provide a compatibility macro that looks something like this: 
```c
#ifndef isgreater
#define isgreater(x,y) ((x) == (x) && (y) == (y) && (x) > (y))
#endif
```
Just my opinion. I'm not a fan of such a macro.

Reply:
I've worked on systems where FP exception traps (for invalid operation, overflow and divide by zero) were enabled by default, so I'm possibly unusually conscious of exceptions. I think most systems don't enable any traps by default though, so you'd probably get away with it in an application like this without raising `SIGFPE`.

Reply:
IIRC, FP exception traps are controlled per application, and since htop uses NaNs extensively, enabling traps would generate a lot of noise. I'm not sure htop's stance on FP exception traps. Would @BenBE answer for this?

Reply:
Yes, it's per-application.  You _shouldn't_ get a lot of noise from using NaNs with traps enabled. The C `NAN` is a "quiet" NaN that just propagate through FP operations.   (IEEE also has signalling NaNs that generate exceptions whenever you touch them, but C doesn't support them).

I think ordering comparisons are the only thing that makes quiet NaNs cause exceptions.

Reply:
We've been using the `isgreater` macro for ordering, thus in this situation I'd likely go with it to properly convey intent. This also solves unlikely cases where that value could accidently be negative.

FWIW: The overall attitude to NaNs in htop is to treat them as acceptable values (if actual value unknown/indeterminate) and regard them as zero if any calculations need to be performed on them. Thus silent handling as far as the source is concerned.

Reply:
This is now being rebased on top of code that's already using `isnan` for the "compressed" value in the same place. Shall I change both to `isgreater`?

Reply:
@kjbracey2 Hello. Not sure if you can test it out, but I believe using `isgreater` can yields smaller and slightly faster code than `isnan`. I think htop's use of `isnan` was a mistake at the start, and if possible, be replaced with `isgreater`, `isless` or such operations.

The technical of `isnan` is that it's a type of `fpclassify` function, and it has special handling for "signaling" NaNs so it nevers throws an FP exception. Htop never uses "signaling" NaNs as far as I knew, and `isnan` would be more expensive than a `x != x` operation. Of course, it would be another issue report for replacing all `isnan` calls in htop with cheaper operations.

Reply:
C99/C11 don't specify anything about signalling NaNs, so that would have surprised me.

I've just checked, and no, GCC and Clang reduce both `isnan` and `isgreater` to a single floating-point comparison.. It's just `x != x` or `x >= 0`. 

https://godbolt.org/z/5P1jGMTdY

Unless you specify `-fsignaling-nans` to GCC, in which case then it makes a special call. That shouldn't be on for htop.

Reply:
Here's the actual code generated in this case:

```
    movsd   (%rax),%xmm0
    ucomisd %xmm1,%xmm1
    jp      skip_add
    addsd   %xmm1,%xmm0
skip_add:
```

Reply:
Ah, I've found the stuff about signaling NaNs in the C extension ISO/IEC TS 18661-1, which does mandate the "no exception for signalling NaNs" for `isnan` and others, but that's not part of the core C standard, which can't produce signaling NaNs and mandates nothing about their behaviour. So special `isnan` SNaN handling isn't on by default in compilers.

Reply:
@kjbracey2 Yes. That's the tricky part of the standards. The IEEE floating point standard defined "signaling" NaNs with a special case that `fpclassify` functions would never throw exceptions on them despite that other operations would throw exceptions on "signaling" NaNs. That's why I add sarcastic quotes on the term "signaling".

Since we can never assume users would set `-fno-signaling-nans` on compiling htop, it might be better for us to avoid the tricky `isnan` operation altogether. Some of the conditionals in htop code would need to check for negative values in addition to NaNs. And for those cases using `isgreater`, `isgreaterequal` etc. would be the future proof way.

Reply:
`-fno-signaling-nans` is the default.

If someone decides to use tougher compiler options that reduce optimisation opportunities, and it gets 0.01% slower as a result, I kind of think it's on them, and I think they'll probably cope. I don't see "this could be slightly slower if people used strange compiler options that serve no purpose for this application" as a very strong argument.

And I'm not a fan of `isgreater(x, 0)` in this bit of code because `if (isgreater(x,0)) { y += x; }` just looks like a weird pseudo-optimisation. "Let's not waste time adding 0". Temptation would be to remove the `if`. Or it could indicate that you are expecting negative values. But the actual intent is just to skip NaNs - yet for some reason we're not using `isnan` like everywhere else.

Having the first and only use of `isgreater` in the code base be to check for NaNs, when there are 50 other places using `isnan` seems a bad idea. Any move away from `isnan` should be codebase-wide, setting a clear new style to follow.  (You could have a local `IS_NAN()` macro that just does `x != x`?)

Reply:
@kjbracey2 When it comes to rendering meter values, there is a `if (isgreater(x,0)) { y += x; }` check as you mentioned. And it's not just for checking NaNs. We are filtering out negative values just in case some meter implementations accidentally put them in.

Yes, I believe replacing `isnan` should be done codebase-wide, but that would be another pull request to be filed.

Reply:
There are no `isgreater`s in `main`. Are you thinking of a pending PR?

The rendering code currently has this:

https://github.com/htop-dev/htop/blob/5558c015f6699c9da47d1cb871318c23dedd4919/Meter.c#L336-L339

Reply:
Since #1271 addressed the issue of `isnan` vs `isgreater` etc., I'm marking this one as resolved.